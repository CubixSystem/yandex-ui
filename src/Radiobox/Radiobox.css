.Radiobox {
    display: inline-block;

    font-family: var(--radiobox-font-family);
    user-select: none;

    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
}

.Radiobox-Radio_disabled {
    pointer-events: none;
}
