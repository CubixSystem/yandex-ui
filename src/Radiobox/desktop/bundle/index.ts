import { compose, composeU } from '@bem-react/core';

import { Radiobox as RadioboxDesktop } from '../../Radiobox@desktop';
// _size
import { withSizeM } from '../../_size/Radiobox_size_m';
import { withSizeS } from '../../_size/Radiobox_size_s';
// _theme
import { withThemeNormal } from '../../_theme/Radiobox_theme_normal@desktop';
import { withThemePseudo } from '../../_theme/Radiobox_theme_pseudo@desktop';
// _view
import { withViewDefault } from '../../_view/Radiobox_view_default@desktop';
import { withViewOutline } from '../../_view/Radiobox_view_outline@desktop';

export * from '../../Radiobox@desktop';

export const Radiobox = compose(
    composeU(withViewDefault, withViewOutline),
    composeU(withSizeM, withSizeS),
    composeU(withThemeNormal, withThemePseudo),
)(RadioboxDesktop);
