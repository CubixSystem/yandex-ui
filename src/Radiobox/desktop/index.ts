export * from '../Radiobox@desktop';
export * from '../_size/Radiobox_size_m';
export * from '../_size/Radiobox_size_s';
export * from '../_theme/Radiobox_theme_normal@desktop';
export * from '../_theme/Radiobox_theme_pseudo@desktop';
export * from '../_view/Radiobox_view_default@desktop';
export * from '../_view/Radiobox_view_outline@desktop';
