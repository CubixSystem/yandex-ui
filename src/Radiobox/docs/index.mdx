---
id: Radiobox
path: /components/radiobox/examples

title: Radiobox
description: Компонент для создания радиопереключателя.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Radiobox
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Radiobox.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fcontrols-radiobox-desktop--playground

tabs:
 - Examples
 - Code
---
