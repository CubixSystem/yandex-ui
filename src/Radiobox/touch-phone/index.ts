export * from '../Radiobox@touch-phone';
export * from '../_size/Radiobox_size_m';
export * from '../_size/Radiobox_size_s';
export * from '../_theme/Radiobox_theme_normal';
export * from '../_theme/Radiobox_theme_pseudo';
export * from '../_view/Radiobox_view_default';
export * from '../_view/Radiobox_view_outline';
