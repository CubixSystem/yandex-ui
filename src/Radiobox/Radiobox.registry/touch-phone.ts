import { Registry } from '@bem-react/di';

import { cnRadiobox } from '../Radiobox';
import { RadioboxRadio } from '../Radio/Radiobox-Radio';

export const radioboxRegistry = new Registry({ id: cnRadiobox() });

radioboxRegistry.set('Radio', RadioboxRadio);
