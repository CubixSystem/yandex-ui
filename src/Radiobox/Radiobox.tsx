import React, { FC, Ref, ReactNode, useState, useEffect, ChangeEventHandler } from 'react';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { IRadioboxRegistry } from './Radiobox.registry';
import { RadioGroupProvider } from './Radiobox.context';
import { RadioboxRadioProps } from './Radio/Radiobox-Radio';
import './Radiobox.css';

export const cnRadiobox = cn('Radiobox');

// Для options используем label вместо children, а так же удаляем
// checked и onChange, т.к. они предполагают использование `Radio` элемента.
type PartialRadioProps = Omit<RadioboxRadioProps, 'children' | 'checked' | 'onChange'>;

export type RadioOptionProps = PartialRadioProps & {
    /**
     * Текст подписи к переключателю
     */
    label: ReactNode;
};

export type RadioboxProps = {
    /**
     * Набор переключателей
     *
     * Может быть использовано вместо `children`
     *
     * @default []
     */
    options?: RadioOptionProps[];

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: Ref<HTMLSpanElement>;

    /**
     * Неактивное состояние всей группы переключателей
     */
    disabled?: boolean;

    /**
     * Набор переключателей c использованием `Radio` элемента
     *
     * Может быть использовано вместо `options`
     */
    children?: ReactNode;

    /**
     * Дополнительный класс у корневого DOM-элемента
     */
    className?: string;

    /**
     * Текущее выбранное значение в группе
     */
    value?: string;

    /**
     * Имя для всех `Radio` элементов
     */
    name?: string;

    /**
     * Коллбек, который срабатывает при изменении значения
     */
    onChange?: ChangeEventHandler<HTMLInputElement>;
};

/**
 * Компонент для создания радиопереключателя.
 * @param {RadioboxProps} props Свойства компонента.
 */
export const Radiobox: FC<RadioboxProps> = ({
    children,
    className,
    disabled,
    innerRef,
    name,
    onChange,
    options = [],
    value,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    // @ts-ignore
    view: _view,
    // @ts-ignore
    size: _size,
    // @ts-ignore
    theme: _theme,
}) => {
    const { Radio } = useComponentRegistry<IRadioboxRegistry>(cnRadiobox());
    const [context, setContext] = useState({ onChange, value, disabled, name });

    useEffect(() => {
        setContext({ onChange, value, disabled, name });
    }, [onChange, value, disabled, name]);

    return (
        <span className={cnRadiobox(null, [className])} ref={innerRef} role="radiogroup">
            {/* prettier-ignore */}
            <RadioGroupProvider value={context}>
                {children || options.map(({ label, ...option }) => (
                    <Radio key={option.value} {...option}>
                        {label}
                    </Radio>
                ))}
            </RadioGroupProvider>
        </span>
    );
};

Radiobox.displayName = cnRadiobox();
