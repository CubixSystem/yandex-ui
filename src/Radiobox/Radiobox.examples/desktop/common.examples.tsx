import React, { useState } from 'react';

import { Radiobox, RadioOptionProps } from '../../Radiobox.bundle/desktop';
import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: createDecorators({ scope: 'desktop' }),
    parameters,
};

export const Size = () => {
    const [value1, setValue1] = useState('a');
    const [value2, setValue2] = useState('a');
    const options: RadioOptionProps[] = [
        { label: 'Option A', value: 'a' },
        { label: 'Option B', value: 'b' },
        { label: 'Option C (disabled)', value: 'c', disabled: true },
    ];

    return (
        <>
            {'_size_m: '}
            <Radiobox
                size="m"
                view="default"
                value={value1}
                onChange={(event) => setValue1(event.target.value)}
                options={options}
            />
            <br />
            {'_size_s: '}
            <Radiobox
                size="s"
                view="default"
                value={value2}
                onChange={(event) => setValue2(event.target.value)}
                options={options}
            />
        </>
    );
};

Size.story = {
    name: 'size',
};

export const Theme = () => {
    const [value1, setValue1] = useState('a');
    const [value2, setValue2] = useState('a');
    const options: RadioOptionProps[] = [
        { label: 'Option A', value: 'a' },
        { label: 'Option B', value: 'b' },
        { label: 'Option C (disabled)', value: 'c', disabled: true },
    ];

    return (
        <>
            {'_theme_normal: '}
            <Radiobox
                size="m"
                theme="normal"
                value={value1}
                onChange={(event) => setValue1(event.target.value)}
                options={options}
            />
            <br />
            {'_theme_pseudo: '}
            <Radiobox
                size="m"
                theme="pseudo"
                value={value2}
                onChange={(event) => setValue2(event.target.value)}
                options={options}
            />
        </>
    );
};

Theme.story = {
    name: 'theme',
};

export const View = () => {
    const [value, setValue] = useState('a');
    const options: RadioOptionProps[] = [
        { label: 'Option A', value: 'a' },
        { label: 'Option B', value: 'b' },
        { label: 'Option C (disabled)', value: 'c', disabled: true },
    ];

    return (
        <>
            {'_view_default: '}
            <Radiobox
                size="m"
                view="default"
                value={value}
                onChange={(event) => setValue(event.target.value)}
                options={options}
            />
        </>
    );
};

View.story = {
    name: 'view',
};
