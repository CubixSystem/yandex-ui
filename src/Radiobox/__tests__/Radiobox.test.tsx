import React, { createRef } from 'react';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { Radiobox as RadioboxCommon } from '../Radiobox';
import { Radiobox as RadioboxDesktop } from '../desktop';
import { Radiobox as RadioboxTouchPad } from '../touch-pad';
import { Radiobox as RadioboxTouchPhone } from '../touch-phone';

const platforms = [
    ['desktop', RadioboxDesktop],
    ['touch-pad', RadioboxTouchPad],
    ['touch-phone', RadioboxTouchPhone],
] as const;

const options = [
    { label: 'Option A', value: 'a' },
    { label: 'Option B', value: 'b' },
];

describe.each(platforms)('Radiobox@%s', (_platform, Radiobox) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Radiobox onChange={() => null} value="a" options={options} className="mix" name="radio" />,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(RadioboxCommon.displayName).toBe('Radiobox');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLElement>();
        render(<Radiobox onChange={() => null} value="" options={[]} innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен вызывать событие onChange с event при клике на контрол', () => {
        // TODO: при переходе на 17 реакт `e.persist` убрать
        const onChange = jest.fn((e) => e.persist());

        render(<Radiobox value="a" options={options} onChange={onChange} />);

        fireEvent.click(screen.getAllByRole('radio')[1]);

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange.mock.calls[0][0].target.value).toBe('b');
    });

    test('должен устанавливать checked/aria-checked при обновлении свойства value', () => {
        const { setProps } = render(<Radiobox onChange={() => null} options={options} value="a" />);

        const [a1, b1] = screen.getAllByRole('radio');
        expect(a1).toBeChecked();
        expect(a1).toHaveAttribute('aria-checked', 'true');
        expect(b1).not.toBeChecked();
        expect(b1).toHaveAttribute('aria-checked', 'false');

        setProps({ value: 'b' });

        const [a2, b2] = screen.getAllByRole('radio');
        expect(a2).not.toBeChecked();
        expect(a2).toHaveAttribute('aria-checked', 'false');
        expect(b2).toBeChecked();
        expect(b2).toHaveAttribute('aria-checked', 'true');
    });

    test('должен добавлять и удалять атрибут disabled при обновлении через props', () => {
        const { setProps } = render(<Radiobox onChange={() => null} value="a" options={options} />);

        const [a1, b1] = screen.getAllByRole('radio');
        expect(a1).not.toBeDisabled();
        expect(b1).not.toBeDisabled();

        setProps({ disabled: true });

        const [a2, b2] = screen.getAllByRole('radio');
        expect(a2).toBeDisabled();
        expect(b2).toBeDisabled();
    });

    test('должен позволять блокировать и разблокировать отдельные опции через props', () => {
        const { setProps } = render(<Radiobox onChange={() => null} value="a" options={options} />);

        const [a1, b1] = screen.getAllByRole('radio');
        expect(a1).not.toBeDisabled();
        expect(b1).not.toBeDisabled();

        setProps({
            options: [
                { label: 'Option A', value: 'a' },
                { label: 'Option B', value: 'b', disabled: true },
            ],
        });

        const [a2, b2] = screen.getAllByRole('radio');
        expect(a2).not.toBeDisabled();
        expect(b2).toBeDisabled();
    });
});
