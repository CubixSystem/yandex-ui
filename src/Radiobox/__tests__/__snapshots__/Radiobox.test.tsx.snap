// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radiobox@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;
